### ISSUE TITLE
Intermittent null reference exception when connecting to EventStore

### ISSUE THREAD

**Issue by Poimen** [EventStore/EventStore#1857]
Hi,

We are experiencing an intermittent null reference exception when using the client 4.1.1.

Our code is used in a threaded environment with each stream assigned its own thread. Each thread attempts to read a particular stream after running the connection code. We protect the connection creation with locks so that only a single connection is opened to EventStore. We are opening a TCP connection with the server.

Conceptually our code does:
```
if(notConnected) {
/* protect code with lock and monitor conditional variables */
Connection = EventStoreConnection.Create(...);
Connection.ConnectAsync().Wait();
/* wait for the connection callback from Connection.Connected to fire */
}
Connection.SubscribeToStreamFrom(...)
```
At the point of calling `SubscribeToStreamFrom` everything appears correct.

For the exception message:
```
  "ExceptionMethod": "8\nSubscribeToStream\nServices.EventStore, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null\nServices.EventStore.Provider\nVoid SubscribeToStream(System.String, System.Nullable`1[System.Int64])",
```

I have validated that the `Connection` object coming back from `EventStoreConnection.Create` is valid before calling `SubscribeToStreamFrom`. So far I can't see anything on our side that would cause this intermittent behaviour.

Any clues on what could be causing this or how to improve debugging this?

Thanks,
Richard

**Comment by jageall** [comment]
Do you have a stacktrace for the exception? That would help. 

Some improvements have been made in this area for v5 of the client which is compatible with v4.1.1. a release candidate (or release) with those fixes in should be out next week.

The most likely cause if it is in the EventStore client code is that the connection dropped in the middle of an operation, but again the stacktrace would help with diagnosing.

**Comment by Poimen** [comment]
So it turns out my debugging code wasn't quite as accurate as I thought. I started looking into the connection dropping and picked up my debug handling wasn't so accurate to catch the correct error.  I have seen intermittent connection drops...however...

The callstack won't really help as the `Connection` (that I thought was fine 🙈) is coming back as `null` - hence null reference exception when we use it. So the callstack just points to our code - which then led to another problem...

At the end of it, if I increased the thread pressure, then the connection would fall over reasonably quickly - which meant that it turned out that the connection was prematurely being returned to the calling code and used (which then goes 💥)


Sorry for the red-herring, but grateful for the quick response.

Thanks,
Richard
